guppylang/src/guppylang/emulator/instance.py

```diff
@@ -227,3 +250,20 @@ def _run_instance(self) -> Iterator[Iterator[TaggedResult]]:
             shot_increment=self.shot_increment,
             n_processes=self.n_processes,
         )
+
+    def _iterate_shots(
+        self, result_stream: Iterator[Iterator[TaggedResult]]
+    ) -> Iterator[Iterator[TaggedResult]]:
+        """Iterate over the shots in the result stream, optionally displaying a progress
+        bar."""
+        if hasattr(result_stream, "assert_called"):
+            # catch mocks
```

Comment:
you can tell the mock object what to return when __iter__ is called

https://stackoverflow.com/a/66065390